File: 0020-valid-parentheses/0020-valid-parentheses.py
from collections import deque

class Solution:
    def isValid(self, s: str) -> bool:
        d = deque()
        characters = {'(': ')', '{': '}', '[': ']'}

        for c in s:
            if c in characters.keys():
                d.append(c)
                continue
            else:
                if len(d) == 0: return False
                opening = d.pop()
                closing = characters[opening]
                if closing != c: return False
    
        if len(d) == 0:
            return True
        else:
            return False

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        if (s.length()%2 != 0){
            return false;
        }

        Stack<Character> st = new Stack<>();
        Map<Character, Character> hashMap = new HashMap<>();
        hashMap.put(')', '(');
        hashMap.put('}', '{');
        hashMap.put(']', '[');

        for (int i = 0; i < s.length(); i++){
            if (s.charAt(i) == '(' || s.charAt(i) == '{' || s.charAt(i) == '['){
                st.push(s.charAt(i));
            } else {
                if (st.isEmpty() || hashMap.get(s.charAt(i)) != st.pop()){
                    return false;
                }
            }
        }

        return (st.isEmpty()) ? true : false;
    }
}